engine/access/subscription/block_tracker.go

```diff
@@ -0,0 +1,150 @@
+package subscription
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/counters"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/state/protocol"
+
+	"github.com/onflow/flow-go/storage"
+)
+
+// BlockTracker is an interface for tracking blocks and handling block-related operations.
+type BlockTracker interface {
+	BaseTracker
+	// GetHighestHeight returns the highest height based on the specified block status.
+	// Only flow.BlockStatusFinalized and flow.BlockStatusSealed are allowed.
+	//
+	// Returns expected errors:
+	// - codes.InvalidArgument: If the specified block status is unsupported.
+	GetHighestHeight(flow.BlockStatus) (uint64, error)
+	// ProcessOnFinalizedBlock drives the subscription logic when a block is finalized.
+	// The input to this callback is treated as trusted. This method should be executed on
+	// `OnFinalizedBlock` notifications from the node-internal consensus instance.
+	ProcessOnFinalizedBlock() error
+}
+
+var _ BlockTracker = (*BlockTrackerImpl)(nil)
+
+// BlockTrackerImpl is an implementation of the BlockTracker interface.
+type BlockTrackerImpl struct {
+	*BaseTrackerImpl
+	state       protocol.State
+	broadcaster *engine.Broadcaster
+
+	// finalizedHighestHeight contains the highest consecutive block height for which we have received a new notification.
+	finalizedHighestHeight counters.StrictMonotonousCounter
+	// sealedHighestHeight contains the highest consecutive block height for which we have received a new notification.
+	sealedHighestHeight counters.StrictMonotonousCounter
+}
+
+// NewBlockTracker creates a new BlockTrackerImpl instance.
+//
+// Parameters:
+// - state: The protocol state used for retrieving block information.
+// - rootHeight: The root block height, serving as the baseline for calculating the start height.
+// - headers: The storage headers for accessing block headers.
+// - highestAvailableFinalizedHeight: The highest available finalized block height.
+// - broadcaster: The engine broadcaster for publishing notifications.
+// - indexReporter: The index reporter for checking indexed block heights.
+// - useIndex: A flag indicating whether to use indexed block heights for validation.
+//
+// Returns:
+// - *BlockTrackerImpl: A new instance of BlockTrackerImpl.
+// - error: An error indicating the result of the operation, if any.
+func NewBlockTracker(
+	state protocol.State,
+	rootHeight uint64,
+	headers storage.Headers,
+	highestAvailableFinalizedHeight uint64,
```

Comment:
not sure we want to pass height here, why can't we query the state the same way we did for sealed? It will be more consistent